### DISCUSSION TITLE
Roland CY-5 not dual triggering

### DISCUSSION THREAD

**Discussion by thijstriemstra** [corrados/edrumulus#35] [General]
I'm testing the Roland CY-5 dual triggering pad with an ESP32 and this config:

| Pad | Type | MIDI note | GPIO |
| --- | --- | --- | --- |
| Roland CY-5 | Hihat | 22, 26, 44 (?) | 32, 26 |

and hitting the pad only produces midi note 26, never 22 (and/or 44, but not sure yet how these are assigned exactly). I'm using the `thijstriemstra_prototype` branch.

I have triple checked the circuit, and since it's a copy of the snare (which properly works, at least with the snare, and i soldered in the same way), I hope it's a software thing and not some thing i screwed up during soldering. I hate hardware bugs.

**Comment by corrados** [comment]
Have you already tried to change the Rim Shot Threshold in edrumuluscontrol.m?

**Reply by thijstriemstra** [reply]
> Have you already tried to change the Rim Shot Threshold in edrumuluscontrol.m?

I will give it a try, but yes I think I already tried changing that parameter. Also, what midi note should I expect for the rim and what note for the rest?

Here are stereo recordings of the pad, where

- A is rim from left to right (where first hit is on that black hard plastic, as well as the last hit)
- B is pad from top to bottom (where first hit is on that back hard plastic in center of pad)

[roland-cy5-pad.zip](https://github.com/corrados/edrumulus/files/7833728/roland-cy5-pad.zip)

**Reply by bzass** [reply]
I've just tried similiar setup but with Teensy. I had to tune pad_settings.rim_shot_threshold as it was indeed not triggering on default settings. I was surprised as I thought that the lower the Threshold, the quicker / more sensitive rim shot is going to be but it seems to be the opposite (I had to raise it to start triggering). I'm experimenting with parameters but for now I used mainly those from CY8:

```
      pad_settings.velocity_threshold   = 20;
      pad_settings.velocity_sensitivity = 25;
      pad_settings.rim_shot_treshold    = 15;
      pad_settings.scan_time_ms         = 6.0f;
      pad_settings.decay_len1_ms        = 10.0f;
      pad_settings.decay_grad_fact1     = 10.0f;
      pad_settings.decay_len2_ms        = 100.0f;
      pad_settings.decay_grad_fact2     = 200.0f;
      pad_settings.decay_len3_ms        = 450.0f;
      pad_settings.decay_grad_fact3     = 30.0f;
```

**Reply by bzass** [reply]
> Have you already tried to change the Rim Shot Threshold in edrumuluscontrol.m?

btw. Octave crashes when I'm trying to launch it (current version from Master branch on Windows)

**Reply by thijstriemstra** [reply]
> Octave crashes when I'm trying to launch it (current version from Master branch on Windows)

I'm using version 6.4.0 (2021-10-30) on windows 10 and hasn't crashed at all so far.

**Reply by thijstriemstra** [reply]
I've tried the latest `Edrumulus::CY5` pad settings, as well as the settings suggested by @bzass, but I can never get any other midi note than 26 out of the pad.

> I had to tune pad_settings.rim_shot_threshold as it was indeed not triggering on default settings.

This gave me some hope but looks like there's a fault in my circuit because it still only triggers one note 😞

**Reply by corrados** [reply]
Have you tried to configure your snare input as a crash input with CY-5 for a test to make sure the cause is really your circuit? Because you know that your snare input is working correctly.

**Reply by thijstriemstra** [reply]
> Have you tried to configure your snare input as a crash input with CY-5 for a test to make sure the cause is really your circuit?

Sounds like a good plan.

> Because you know that your snare input is working correctly.

I'll first have to get 2 midi notes out of the snare again..

**Reply by thijstriemstra** [reply]
When I change pad 0 (snare before) to cy-5 and enable the rimshot:

```c++
  edrumulus.set_pad_type            ( 0, Edrumulus::CY5);
  edrumulus.set_rim_shot_is_used    ( 0, true);
```

I have to change the rimshot treshold to the max (31) and then I can hear the rimshot sound when I hit the hard black plastic in the center of the pad, but never anywhere on the soft rubber of the pad. This means the pad works, circuit works (like it does with snare) but the rimshot treshold is still way too high. I dunno.

This afternoon, I completely resoldered a new version of the circuit and unfortunately the situation is exactly the same (so it wasn't the circuit imo..) 

Does this look right to you?

```c++
edrumulus.set_midi_notes       ( hihat_pad_idx, 22, 22 );
  edrumulus.set_midi_notes_open  ( hihat_pad_idx, 26, 26 );
  edrumulus.set_midi_notes       ( hihatctrl_pad_idx, 44, 44 ); // Hi-Hat pedal hit
  edrumulus.set_midi_ctrl_ch     ( hihatctrl_pad_idx, 4 ); // Hi-Hat control
```

I keep thinking it's a software issue. I can see note 26 but note 22 is never touched. It's very frustrating because this is the last thing I have to fix before I can start using the machine (besides the FD-8 paddle which I still need to test).

I've also never seen midi note 44, a pedal hit fly by in hairless midi. Should that work as well?

**Reply by corrados** [reply]
> Does this look right to you?

This may be your issue. You need to set different MIDI notes, not the same. The first MIDI note (e.g. the first 22) is for the middle of the pad and the second MIDI note (in your case also 22) is for the rim. If you set both to the same value, you will always get the same MIDI not no matter if a rim shot is played or not.

**Reply by corrados** [reply]
> When I change pad 0 (snare before) to cy-5 and enable the rimshot:

Please note that if you set the pad type to CY-5, Edrumulus assumes a rim-switch. If you then connect your DIABOLO pad which has a rim piezo, it will most likely not work correctly.

**Reply by thijstriemstra** [reply]
> The first MIDI note (e.g. the first 22) is for the middle of the pad and the second MIDI note (in your case also 22) is for the rim. If you set both to the same value, you will always get the same MIDI not no matter if a rim shot is played or not.

I see. And what is `set_midi_notes_open  ( hihat_pad_idx, 26, 26 );` used for? This is where my confusion comes from..

And while we're at it, when can i expect 44 to trigger? Is this code statement correct btw?

```c++
edrumulus.set_midi_notes       ( hihatctrl_pad_idx, 44, 44 );
```

> Please note that if you set the pad type to CY-5, Edrumulus assumes a rim-switch. If you then connect your DIABOLO pad which has a rim piezo, it will most likely not work correctly.

I noticed this as well.

**Reply by corrados** [reply]
On the Git main branch I have, e.g., `edrumulus.set_midi_notes      ( 0, 38, 40 ); // snare`
This sets the normal strike on the mesh head to MIDI note 38 and the rim shot to MIDI note 40. If a pad is a mono pad, I usually set both values to the same MIDI note.

**Reply by thijstriemstra** [reply]
Making progress!

With these settings for the hihat:

```c++
  // not sure what difference between set_midi_notes and set_midi_notes_open is.
  // only set_midi_notes_open seems to work?
  //edrumulus.set_midi_notes       ( hihat_pad_idx, 22, 26 );
  edrumulus.set_midi_notes_open  ( hihat_pad_idx, 28, 32 );
  edrumulus.set_midi_notes       ( hihatctrl_pad_idx, 44, 44 ); // Hi-Hat pedal hit
  edrumulus.set_midi_ctrl_ch     ( hihatctrl_pad_idx, 4 ); // Hi-Hat control
```
and these CY-5 settings:

```c++
// HI-HAT (using rim switch, Roland CY-5)
  edrumulus.set_pad_type             ( hihat_pad_idx, Edrumulus::CY5 );
  edrumulus.set_rim_shot_is_used     ( hihat_pad_idx, true );
  edrumulus.set_rim_shot_treshold    ( hihat_pad_idx, 31 );
  edrumulus.set_velocity_threshold   ( hihat_pad_idx, 14 );
  edrumulus.set_velocity_sensitivity ( hihat_pad_idx, 8 );
```

I get note 28 everywhere on the pad, but when I squeeze the side of the pad (choke?), I get note 32 quite consistently.. But I would like to only hear 32 on the side and rest should be 28.

Video:


https://user-images.githubusercontent.com/305679/149023433-75c73f2c-991a-4e8c-a96b-3bbd6594dea0.mp4

**Reply by corrados** [reply]
If you use higher values for "`edrumulus.set_rim_shot_treshold    ( hihat_pad_idx, 31 );`" like, e.g., 50, does this help?

**Reply by corrados** [reply]
I just saw that the rim shot threshold for the CY-8 is also set to a high value: https://github.com/corrados/edrumulus/blob/c76061d1e709c1164da5c6adc2fca65d9c473278/edrumulus_parameters.cpp#L190
So, it seems, I need to adjust the value range. But I'll wait for your feedback which value outside of the 0..31 range works for you.

**Reply by thijstriemstra** [reply]
> If you use higher values for "edrumulus.set_rim_shot_treshold ( hihat_pad_idx, 31 );" like, e.g., 50, does this help?

@corrados it looks like it does because now I consistently see note 32 wherever I hit, without having to squeeze/choke the pad like before. Next step for me, how do I get the other note back though, and only hear 32 on the rim..

**Reply by corrados** [reply]
Some value between 31 and 50 is then your optimal value.

**Reply by thijstriemstra** [reply]
Setting it to `edrumulus.set_rim_shot_treshold    ( hihat_pad_idx, 32 );`  causes only note 32 (rim) to be heard across the whole pad.

Setting it to `edrumulus.set_rim_shot_treshold    ( hihat_pad_idx, 33 );` (or anything larger than 33) prints continous stream of pressure notes in hairless midi (same as if I would squeeze the pad) and only note 32 rim can be played:

![midi-pressure-cy-5](https://user-images.githubusercontent.com/305679/149209192-0a7c1565-f357-46c1-8d13-f2d49c6bf7b9.JPG)

Anything below 32, e.g. `edrumulus.set_rim_shot_treshold    ( hihat_pad_idx, 31 );` results in an inconsistent rim sound, sometimes it's played, sometimes it's not, and most of the time it does both note 26 and 32 instead of just 32.

So not sure what to do next..

**Reply by corrados** [reply]
Same with me. I also do not have an explanation for that behavior. Maybe the detection algorithm must be modified for that type of pad.

**Reply by corrados** [reply]
I just tested it with my CY-8 using my Teensy prototype and it works perfectly for me.

**Reply by corrados** [reply]
Maybe there is a possibility to improve the code to get your problem fixes. I just saw in the code that the settings parameter is multiplied with a factor 9:
https://github.com/corrados/edrumulus/blob/c2f6fde37cf3522dba26efc772190b46a8f46bb4/edrumulus.cpp#L330
You could try to hard-code some test values for rim_switch_treshold like `rim_switch_treshold      = -ADC_MAX_NOISE_AMPL - x` where x = {0, 1, 2, ..., 9}.

**Reply by thijstriemstra** [reply]
Thanks for the feedback. I will give these values a try. To re-iterate the issue:

I am not able to 'widen' the range of the rimshot, it's only triggered on a very tiny area of the pad (area labeled A in image below). I can only either place:

- rimshot note on A + B, so everywhere
- pad note on A + B, so everywhere
- rimshot on A (tiny area), pad on B

For now I settled on the 2nd option and pretend the pad doesn't dual trigger. The use of the FD-8 hihat foot pedal makes up for this loss somewhat. I'm still amazed that the difference between CY-8 and CY-5 is so big.. ps. i bought the cy-5 brandnew, few months ago, even opened it up to check the wiring but all looks good.

![cy-5](https://user-images.githubusercontent.com/305679/150026152-2bf32528-c8cd-4a82-8175-6a6416df6e2b.png)

**Reply by thijstriemstra** [reply]
Here are some tests I did (unfortunately with no success). I used a tom sound for the rim, hihat for the other, so it's easier to hear it trigger.

```cpp
  edrumulus.set_rim_shot_is_used     ( hihat_pad_idx, true );
  edrumulus.set_rim_shot_treshold    ( hihat_pad_idx, 16 ); // default is 12
  edrumulus.set_velocity_threshold   ( hihat_pad_idx, 15 );
  edrumulus.set_velocity_sensitivity ( hihat_pad_idx, 16 );

rim_switch_treshold      = -ADC_MAX_NOISE_AMPL + 1 * ( pad_settings.rim_shot_treshold - 31 );
```

https://user-images.githubusercontent.com/305679/150162170-5c59d549-63b2-4d70-8b31-4aaa982d62ab.mp4

---

```c++
  edrumulus.set_rim_shot_is_used     ( hihat_pad_idx, true );
  edrumulus.set_rim_shot_treshold    ( hihat_pad_idx, 25 ); // default is 12
  edrumulus.set_velocity_threshold   ( hihat_pad_idx, 15 );
  edrumulus.set_velocity_sensitivity ( hihat_pad_idx, 16 );

rim_switch_treshold      = -ADC_MAX_NOISE_AMPL + 1 * ( pad_settings.rim_shot_treshold - 31 );
```

https://user-images.githubusercontent.com/305679/150162565-cf4e180e-2ed3-42b7-b24c-f6f90405746d.mp4

---

```c++
  edrumulus.set_rim_shot_is_used     ( hihat_pad_idx, true );
  edrumulus.set_rim_shot_treshold    ( hihat_pad_idx, 25 ); // default is 12
  edrumulus.set_velocity_threshold   ( hihat_pad_idx, 15 );
  edrumulus.set_velocity_sensitivity ( hihat_pad_idx, 16 );

rim_switch_treshold      = -ADC_MAX_NOISE_AMPL + 4 * ( pad_settings.rim_shot_treshold - 31 );
```

https://user-images.githubusercontent.com/305679/150162893-3da98a49-8688-4dd5-94c2-fd89562afee2.mp4

---

```c++
  edrumulus.set_rim_shot_is_used     ( hihat_pad_idx, true );
  edrumulus.set_rim_shot_treshold    ( hihat_pad_idx, 31 ); // default is 12
  edrumulus.set_velocity_threshold   ( hihat_pad_idx, 15 );
  edrumulus.set_velocity_sensitivity ( hihat_pad_idx, 16 );

rim_switch_treshold      = -ADC_MAX_NOISE_AMPL + 4 * ( pad_settings.rim_shot_treshold - 31 );
```

https://user-images.githubusercontent.com/305679/150163188-8d0efb1b-f31e-424b-8b83-732d2dc784b6.mp4

---

```c++
  edrumulus.set_rim_shot_is_used     ( hihat_pad_idx, true );
  edrumulus.set_rim_shot_treshold    ( hihat_pad_idx, 31 ); // default is 12
  edrumulus.set_velocity_threshold   ( hihat_pad_idx, 15 );
  edrumulus.set_velocity_sensitivity ( hihat_pad_idx, 16 );

  rim_switch_treshold      = -ADC_MAX_NOISE_AMPL + 7 * ( pad_settings.rim_shot_treshold - 31 );
```


https://user-images.githubusercontent.com/305679/150163521-3a952219-ba1f-4a2f-9558-e6a0cc629e1f.mp4

**Reply by corrados** [reply]
Now I can see what your issue is:
![grafik](https://user-images.githubusercontent.com/46655886/150166653-35387f24-2ebb-4599-b504-16b758194a42.png)
Roland cymbals are designed to only output the rim signal if you strike the pad with the "Edge Strike". If you strike the pad at the edge but from above, this is still a "Bow Strike" and should not trigger the rim.

**Reply by corrados** [reply]
> I see. I double-checked the manual and it also mentions this:
> cy-5-manual
> I'll see if I can tune the parameters correctly for this setup.

Sorry, I had to delete your post to avoid copyright issues with the Roland manual (that is actually the reason I painted the figure by hand :-) ).
